### PR TITLE
Refactor downloads to be more robust

### PR DESCRIPTION
**Pull request by nebfield**
Historical approach:

* We had weird intermittent HTTPS problems originally so we switched to FTP downloads
* But the FTP is shared across the EBI and it's often busy (error 421)

New approach:

* HTTPS download preferred, retried up to 5 times
* FTP fallback attempted, retried up to 5 times

The downloads module started off simple but got a bit untidy as we added features, so I've tried to simplify the structure and remove redundant functions.

The main motivation for this PR is making pgsc_calc less likely to throw errors on the INTERVENE platform when the FTP is busy. 

To do:

* checksums ✅
* set local download directory ✅
* include pgsc_calc version number in UA ✅
* fix tests ✅

### CODE REVIEW

**Comment by smlmbrt** [comment]
I marked this for @ens-lgil to review, as he has more expertise on the API & checksums

File: pgscatalog_utils/download/ScoringFileDownloader.py [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+        for scoring_file in scoring_files:
+            download_file(scoring_file.url, scoring_file.local_path, overwrite=self.overwrite,
+                          ftp_fallback=self.ftp_fallback)
+            checksum: ScoringFileChecksum = ScoringFileChecksum.from_scoring_file(scoring_file)
+
+            if not checksum.matches:
+                logger.warning(f"Scoring file {scoring_file.local_path} fails validation")
+                logger.warning(f"Remote checksum: {checksum.remote_checksum}")
+                logger.warning(f"Local checksum: {checksum.local_checksum}")
+                max_attempt = 3
```
Comment (ens-lgil) [inline pgscatalog_utils/download/ScoringFileDownloader.py:45]: Trivial but should we set this variable somewhere else, e.g. in the config module for an easier access ?

File: pgscatalog_utils/download/download_file.py [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+            logger.warning("Attempting FTP fallback")
+            _ftp_fallback_download(url=url, local_path=local_path)
+        else:
+            raise Exception(f"Can't download {url} using HTTPS")
+
+
+def _ftp_fallback_download(url: str, local_path: str) -> None:
+    url = url.replace("https://", "ftp://")
+    max_retries = 5
+    wait_time = 30
```
Comment (ens-lgil) [inline pgscatalog_utils/download/download_file.py:53]: Also store the waiting time somewhere else ?

File: pgscatalog_utils/download/download_scorefile.py [hunk 3]
```diff
@@ -1,185 +1,111 @@
...
+            for term in args.efo:
+                results.append(CatalogQuery(CatalogCategory.TRAIT, term, include_children=True,
+                                            pgsc_calc_version=config.PGSC_CALC_VERSION).get())
         else:
             logger.debug("--trait set, querying traits")
-        pgs_lst = pgs_lst + [query_trait(x, pgsc_calc_info, args.efo_include_children) for x in args.efo]
-
+            for term in args.efo:
+                results.append(CatalogQuery(CatalogCategory.TRAIT, term, include_children=False,
+                                            pgsc_calc_version=config.PGSC_CALC_VERSION).get())
```
Comment (ens-lgil) [inline pgscatalog_utils/download/download_scorefile.py:64]: Seems a bit redundant: could we set a temp variable for the `include_children` boolean, e.g.:
```
        inc_child = False
        if args.efo_include_children:
            logger.debug("--trait set, querying traits (including PGS for child terms)")
            inc_child = True
        else:
            logger.debug("--trait set, querying traits")
        results.append(CatalogQuery(CatalogCategory.TRAIT, term, include_children=inc_child,
                                  pgsc_calc_version=config.PGSC_CALC_VERSION).get())

```

**Comment by nebfield** [comment]
Thanks for the review @ens-lgil  🤠

**Review by ens-lgil — APPROVED** [review]

**Pull request merged**